Time Ranges: getting nicely ahead of ourselves (#166)
andycroll/andycroll.com: _posts/2022/2022-10-17-time-ranges-are-clearer-when-querying-databases.md

---
title: "Time Ranges are Clearer When Querying Databases"
description: "Avoid text input to Active Record queries where you can"
layout: article
category: ruby
image:
  base: "2022/time-ranges-are-clearer-when-querying-databases"
  alt: "Alarm Clock on pink and blue"
  credit: "Icons8 Team"
  source: "https://unsplash.com/photos/dhZtNlvNE8M"
---

> Active Record‘s query interface has _many_ different ways to turn your logic into SQL to get rows from your database.

Reviewer: Not clear why you'd need to "turn your logic into SQL".

Reviewer: Ooh — now I'm at the examples I know what you mean: so that you can copy and paste into your DB, right?

> I extolled the virtues of the [core extensions to `Date` and `Time`](/ruby/date-and-time-extensions-are-an-unsung-hero-of-active-support) and in our web applications we‘re often extracting records from databases where time periods are on of our primary filters. So let’s use those extensions to query our database.

Reviewer: *I _previously_ extolled...

*are _one_

Author: ```suggestion
I previously extolled the virtues of the [core extensions to `Date` and `Time`](/ruby/date-and-time-extensions-are-an-unsung-hero-of-active-support) and in our web applications we‘re often extracting records from databases where time periods are one of our primary filters. So let’s use those extensions to query our database.
```

## Instead of...

> ...using a text-based SQL fragments to query time fields.

Reviewer: *:

Author: ```suggestion
...using a text-based SQL fragments to query time fields:
```

```ruby
User.where("created_at > ? AND created_at < ?", 2.weeks.ago, 1.week.ago).to_sql
#=> "SELECT ... WHERE (created_at > '2022-08-31 11:29:53.945995' AND created_at < '2022-09-07 11:29:53.946280')"
User.where("created_at > ? AND created_at < ?", 2.weeks.ago.beginning_of_day, 2.weeks.ago.end_of_day).to_sql
#=> "SELECT ... WHERE (created_at > '2022-08-31 00:00:00' AND created_at < '2022-08-31 23:59:59.999999')"
User.where("created_at >= ?", 2.weeks.ago.beginning_of_day).to_sql
#=> "SELECT ... WHERE \"users\".\"created_at\" >= '2022-08-31 00:00:00'"
```

## Use...

...`Range`s of time passed as an argument.

```ruby
User.where(created_at: (2.weeks.ago..1.week.ago)).to_sql
#=> "SELECT ... WHERE \"users\".\"created_at\" BETWEEN '2022-08-31 11:29:33.248193' AND '2022-09-07 11:29:33.248938'"
User.where(created_at: 2.weeks.ago.all_day).to_sql
#=> "SELECT ... WHERE \"users\".\"created_at\" BETWEEN '2022-08-31 00:00:00' AND '2022-08-31 23:59:59.999999'"
User.where(created_at: (2.weeks.ago.beginning_of_day..)).to_sql
#=> "SELECT ... WHERE \"users\".\"created_at\" >= '2022-08-31 00:00:00'"
```

## Why?

The syntax in this case is shorter and clearer.

> You're also likely to make less mistakes as the SQL statement generated by the scope won‘t have a big splurge of your text inserted into it.

Reviewer: *fewer

> In addition the SQL generated by the framework when passed a range takes advantage of the correct `BETWEEN` SQL syntax.

Reviewer: Why does this matter?

## Why not?

> It seems that passing `Range`s to Active Record scopes seems less common than it should be. I presume this is because `String`-based conditions are listed first in the guides documentation, and were—for early versions of Rails—the primary syntax.

Reviewer: Repetition of "seems" is awkward.

Doesn't really explain "Why not?".

Author: ```suggestion
Passing `Range`s to Active Record scopes seems less common than it should be. I presume this is because `String`-based conditions are listed first in the guides documentation, and were—for early versions of Rails—the primary syntax.

There’s very little reason not to use this more modern syntax. It's clearer to read and reason about when you’re using the features of Active Record than throwing `String`s around. That way it stands out more when you _do_ have to drop down to `String`s for more complicated SQL queries.
```

> My final example uses an endless `Range` which was introduced in Ruby 2.6 (in 2018) and it took a little while to include into Rails so you might have seen

Reviewer: This post is also unfinished...??

Author: This is so weird. I must have somehow committed in progress versions.
